docs/plugins.md

```diff
@@ -0,0 +1,133 @@
+      usage       'foo [options]'
+      summary     '3scale foo'
+      description '3scale foo command'
+      flag :h, :help, 'show help for this command' do |_, cmd|
+        puts cmd.help
+        exit 0
+      end
+      run do |opts, args, _|
+        puts "foo done"
+      end
+    end
+  end
+end
+ThreeScaleToolbox::CLI.add_command(FooCommand)
+
+$ RUBYOPT=-Ilib 3scale foo
+Hello World
+```
+Few things worth to highlight.
+- Your module must be extended by *ThreeScaleToolbox::Command* module. It allows your command to be added to CLI command tree.
+- Must implement `command` module function and return instance of `Cri::Command` from [cri](https://github.com/ddfreyne/cri)
+- Add your command to `3scale` CLI command tree by calling `ThreeScaleToolbox::CLI.add_command`
+
+Your plugin help is also available using builtin *help* command
+
+```
+$ RUBYOPT=-Ilib 3scale help foo
+NAME
+    foo - foo command
+
+USAGE
+    3scale foo [options]
+
+DESCRIPTION
+    This command does a lot of stuff.
+
+OPTIONS
+    -h --help         show help for this command
+
+OPTIONS FOR 3SCALE
+    -v --version      Prints the version of this command
+```
+
+Let's create a plugin to add a `simple hello world` subcommand for the builtin *copy* command.
+
+```
+$ cat lib/3scale_toolbox_plugin.rb
+require '3scale_toolbox/base_command'
+require '3scale_toolbox/commands/copy_command'
+
+module FooCommand
+  extend ThreeScaleToolbox::Command
+
+  def self.command
+    Cri::Command.define do
+      name        'foo'
+      usage       'foo [options]'
+      summary     '3scale copy foo'
+      description '3scale copy foo subcommand'
+      flag :h, :help, 'show help for this command' do |_, cmd|
+        puts cmd.help
+        exit 0
+      end
+      run do |opts, args, _|
+        puts "foo done"
+      end
+    end
+  end
+end
+
+ThreeScaleToolbox::Commands::CopyCommand.add_subcommand(FooCommand)
+
+$ RUBYOPT=-Ilib 3scale copy foo
+foo done
+```
+
+Few things worth to highlight.
+- Your module must be extended by *ThreeScaleToolbox::Command* module. It allows your command to be added to CLI command tree.
+- Must implement `command` module function and return instance of `Cri::Command` from [cri](https://github.com/ddfreyne/cri)
+- Add your subcommand to `3scale` CLI command tree by calling parent command's module's `add_subcommand` method.
+
+Checking `copy` command help, it can be verified the new subcommand `foo` is added.
+
+```
+$ RUBYOPT=-Ilib 3scale help copy
+NAME
+    copy - 3scale CLI copy
+
+USAGE
+    3scale copy <command> [options]
+
+DESCRIPTION
+    3scale CLI copy tools to manage your API from the terminal.
+
+SUBCOMMANDS
+    foo         3scale copy foo
+    service     3scale CLI copy service
+
+OPTIONS
+    -h --help         show help for this command
+
+OPTIONS FOR 3SCALE
+    -v --version      Prints the version of this command
+```
+
+Now, package your plugin as a [gem](https://guides.rubygems.org/make-your-own-gem/) and let us know about it.
```

Comment:
I think this part is the most fragile. Gems have a flat namespace where you basically would need to name your plugin as `3scale_toolbox-myplugin` to avoid polluting this global shared namespace. I'd advocate having the code installed in something such as `/usr/share/3scale_toolbox/plugins` (or `.local/share`...) or having a config file setting to specify it.

At the very least, here we should suggest using a specific prefix for the gem names, and possibly enforce it when looking up plugin code.

Update: I see we are using a specific path to look up plugins, but I guess that is still not enforcing things like the above, and it is not clear how to install such plugins from this doc.

Reply:
That is a common practice in Ruby ecosystem. https://guides.rubygems.org/name-your-gem/

We just require gems to have `3scale_toolbox_plugin.rb` to be automatically loaded by the `3scale` executable. 

I don't really see any need to enforce any rules. People are free to do whatever they want to do.
It is common practice to name extensions prefixed with the gem name. Also it is common practice to namespace all your code with the correct namespace according to your gem name.